docs/promo-codes/index.md

```diff
@@ -0,0 +1,462 @@
+
+<<< Discount.re#get-free-burgers{13-14,21}
+
+This is a straightforward refactor, but to fix your tests, you would need to
+reproduce the error messages verbatim in `DiscountTests`, e.g.:
+
+```reason
+test("0 burgers, no discount", () =>
+  expect
+  |> equal(
+        Discount.getFreeBurgers([
+          Hotdog,
+          Sandwich(Ham),
+          Sandwich(Turducken),
+        ]),
+        Error("To enjoy this promo, buy at least 2 burgers"),
+      )
+);
+```
+
+You would need to change your tests every time the error message changes, which
+is annoying. But error messages don't necessarily need to be inside
+`Discount.getFreeBurgers`, there should be a way to indicate what an error is
+without using the actual error message.
+
+## Polymorphic variants
+
+Instead of using a string as the argument for the `Error` constructor, you can
+instead use a *polymorphic variant*:
+
+<<< Discount.re#get-free-burgers-poly{13-14}
+
+Polymorphic variants are similar to the variants you've seen before, the main
+difference being that you don't need to declare the constructors beforehand.
+Instead, you can freely use polymorphic variant constructors inside a function,
+and the type of the function is inferred from the usage. For example, the type
+of `Discount.getFreeBurgers` is now
+
+```reason
+list(Item.t) => result(float, [> `NeedOneBurger | `NeedTwoBurgers ])
+```
+
+::: tip
+
+The name of a polymorphic variant constructor must start with the backtick (`` `
+``) character.
+
+:::
+
+Refactor `Discount.getHalfOff` to return `result` as well[^2]:
+
+```reason
+switch (meetsCondition) {
+| false => None // [!code --]
+| false => Error(`NeedMegaBurger) // [!code ++]
+| true =>
+  let total =
+    items
+    |> ListLabels.fold_left(~init=0.0, ~f=(total, item) =>
+          total +. Item.toPrice(item)
+        );
+  Some(total /. 2.0); // [!code --]
+  Ok(total /. 2.0); // [!code ++]
+```
+
+The type of `Discount.getHalfOff` is now
+
+```reason
+list(Item.t) => result(float, [> `NeedMegaBurger ])
+```
+
+## Fixing the tests
+
+Fixing the tests is mostly a mechanical and consists of these steps:
+
+- Replace `Some` with `Ok`
+- Replace `None` with ``Error(`SomeConstructor)``
+- Replace `equal` with `deepEqual`
+
+However, there is a little wrinkle. What if you misspell one of the polymorphic
+variant constructors?
+
+```reason{5}
+test("1 burger, no discount", () =>
+  expect
+  |> deepEqual(
+       Discount.getFreeBurgers([Hotdog, Sandwich(Ham), Burger(burger)]),
+       Error(`NeedOneBurgers),
+     );
+```
+
+The constructor should be called `` `NeedOneBurger`` but here it's been
+misspelled as `` `NeedOneBurgers``, yet it still compiles! This is due to the
+open-ended nature of polymorphic variants. The compiler knows that
+`Discount.getFreeBurgers` currently can't return ``Error(`NeedOneBurgers)``, but
+because the constructors for polymorphic variants don't need to be defined
+up-front, it has no way to know that the function will **never** return
```

Comment:
Maybe regular variants should be used here? It's unclear to me (as a reader) why polyvars are used as arguments of `Error`. Edit: ok, i see there'll be more about this in next chapter.